src/lib/utils.ts

```diff
@@ -87,32 +84,102 @@ export const promiseErrorMultiplier = <T, Q extends Error>(
   });
 };
 
+export const generateGatewayMap = async (
+  l2Multicaller: MultiCaller,
+  l2Network: L2Network,
+  l1Provider: Provider
+) => {
+  const fromBlock = 0;
+  const toBlock = await l1Provider.getBlockNumber();
+  const topic0 = ethers.utils.id('GatewaySet(address,address)');
```

Comment:
An edge case we want to check for is that the Gateway has been set on the L1 side but not the L2 side (either because the retryable ticket failed or just hasn't been executed on L2 yet); we should make sure there's a corresponding GatewaySet event emitted at the L2GatewayRouter for each L1 event as a precondition for including it in our map.